### PR TITLE
[Remove] Add confirmation dialog

### LINKED ISSUE #563: [Remove] Add confirmation dialog
<p>Add warning dialog when the remove action is performed</p> <p>See UI docs attached to <a href="https://trunk.arc.nasa.gov/jira/browse/WTD-846" title="[Timeline] First implementation of new timeline design" class="issue-link" data-issue-key="WTD-846"><del>WTD-846</del></a>, p. 99</p>

Migrated from: https://trunk.arc.nasa.gov/jira/browse/WTD-935
Originally created: Wed, 25 Feb 2015 13:16:32 -0800
Updated: Thu, 14 Jan 2016 16:41:44 -0800

### PR DESCRIPTION
**Pull request by tobiasbrown**
I've added a blocking message dialog which prompts the user to confirm the remove action.  This dialog associates the existing remove function (`removeFromContext`) with a button's callback.  This has broken existing test cases and I'm wondering how these can be preserved.  Are we able to mock button selection on a call to `showBlockingMessage` or is it best to implement this dialog with a call to `getUserChoice` which returns a Promise?

I did experiment with `getUserChoice` but ran into a few little issues.  The `overlay-options` template references different property names to the ones defined by the `DialogModel` type.

## Author Checklist
1. Changes address original issue? Yes
2. Unit tests included and/or updated with changes? WIP (see comments above)
3. Command line build passes? No
4. Changes have been smoke-tested? Yes

Closes #563

### CODE REVIEW

File: platform/commonUI/edit/src/actions/RemoveDialog.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    {
+                        label: 'Cancel',
+                        callback: function () {
+                            dialog.dismiss();
+                        }
+                    }
+                ]
+            };
+
+        dialog = this.dialogService.showBlockingMessage(model);
```
[the remark this answers is not in the document]
Comment (tobiasbrown) [reply platform/commonUI/edit/src/actions/RemoveDialog.js]: Thanks for the guidance about how to preserve those test cases.  I had to return the created `RemoveDialog` from `RemoveAction.perform` to invoke `removeCallback`.  In the end, I didn't mock `showBlockingMessage` so I hope that's ok.  Do you need an updated author checklist?

**Comment by charlesh88** [comment]
@akhenry Can we get this reviewed and pushed through? I would like this in place prior to upcoming changes to be made for #1978.

**Comment by akhenry** [comment]
@tobiasbrown My apologies for how long it's taken to get back to you on this. This PR looks really great, good job.

I'm not so sure we should be returning the dialog from the `perform` function on `RemoveAction`. Semantically, it's not a natural consequence of the action itself. While testability is a useful guide in code design, I don't think we should have code specifically to support tests.

An alternative approach here is to spy on `DialogService.showBlockingMessage()` and get the `model` object from that. We can then use `model.primaryOption.callback()`to invoke the `removeCallback` without having to return the dialog from `RemoveAction`.

**Comment by tobiasbrown** [comment]
Thank you for your feedback @akhenry.  I've made some updates so that `RemoveAction.perform()` no longer returns the dialog or any value.  In the tests, the remove callback is triggered after spying on `DialogService.showBlockingMessage()` as you suggested.

**Comment by kundanvishen** [comment]
@akhenry Can you please review this PR?
If it does not solve the issue, I would like to jump in and help.

File: platform/commonUI/edit/src/actions/RemoveAction.js [hunk 2]
```diff
@@ -23,111 +23,121 @@
...
+    function RemoveAction(dialogService, navigationService, context) {
+        this.domainObject = (context || {}).domainObject;
+        this.dialogService = dialogService;
+        this.navigationService = navigationService;
+    }
+
+    /**
+     * Perform this action.
+     *
+     * @returns {RemoveDialog}
```
Comment (larkin) [inline platform/commonUI/edit/src/actions/RemoveAction.js]: @tobiasbrown JSdoc needs to be updated to reflect no return value.

Comment (tobiasbrown) [reply platform/commonUI/edit/src/actions/RemoveAction.js]: Thanks @larkin.  I've removed the `@returns` tag.

**Comment by akhenry** [comment]
This looks great. Thanks for your effort here @tobiasbrown !

We have updated to Jasmine 3 in the time since this PR was first pushed, so I have updated from master and fixed the test spec to be compliant.

## Reviewer Checklist

* Changes appear to address issue? Y
* Appropriate unit tests included? Y
* Code style and in-line documentation are appropriate? Y
* Commit messages meet standards? Y

**Pull request merged**